rm-portal/src/main/java/org/ow2/proactive_grid_cloud_portal/rm/server/ExportNodeSourceServlet.java

```diff
@@ -0,0 +1,123 @@
+import java.util.Iterator;
+import java.util.List;
+
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+import org.apache.commons.fileupload.FileItem;
+import org.apache.commons.fileupload.FileUploadException;
+import org.apache.commons.fileupload.disk.DiskFileItemFactory;
+import org.apache.commons.fileupload.servlet.ServletFileUpload;
+import org.codehaus.jettison.json.JSONException;
+import org.codehaus.jettison.json.JSONObject;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+/**
+ * Servlet invoked when a node source is exported to a file. The request
+ * content is expected to be in JSON format, containing the configuration of a
+ * node source.
+ */
+public class ExportNodeSourceServlet extends HttpServlet {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(ExportNodeSourceServlet.class);
+
+    private static final String NODE_SOURCE_FILE_NAME_SUFFIX = "-configuration.json";
+
+    public static final String SERVLET_MAPPING = "exportnodesource";
+
+    public static final String MAIN_FORM_ITEM_NAME = "nodeSourceJson";
+
+    public static final String NODE_SOURCE_NAME_KEY = "nodeSourceName";
+
+    @Override
+    protected void doPost(HttpServletRequest request, HttpServletResponse response) {
+        downloadJsonFile(request, response);
+    }
+
+    @Override
+    protected void doGet(HttpServletRequest request, HttpServletResponse response) {
+        downloadJsonFile(request, response);
+    }
+
+    private void downloadJsonFile(HttpServletRequest request, HttpServletResponse response) {
+        try {
+            response.setContentType("application/json");
+            String jsonContent = extractJsonStringFromRequest(request);
+            String nodeSourceName = extractNodeSourceName(jsonContent);
+
+            response.setHeader("Content-disposition",
+                               "attachment; filename=" + nodeSourceName + NODE_SOURCE_FILE_NAME_SUFFIX);
+            response.setHeader("Location", nodeSourceName + NODE_SOURCE_FILE_NAME_SUFFIX);
+            response.getWriter().write(jsonContent);
+        } catch (Throwable t) {
+            try {
+                LOGGER.warn("Export node source failed", t);
+                response.getWriter().write(t.getMessage());
+            } catch (IOException e) {
+                LOGGER.warn("Failed to return node source export error to client, error was:" + t.getMessage(), e);
+            }
+        }
+    }
+
+    private String extractJsonStringFromRequest(HttpServletRequest request) throws FileUploadException {
+        DiskFileItemFactory factory = new DiskFileItemFactory();
+        factory.setSizeThreshold(4096);
+        factory.setRepository(new File(System.getProperty("java.io.tmpdir")));
+        ServletFileUpload upload = new ServletFileUpload(factory);
+        upload.setSizeMax(1000000);
+        List<?> fileItems = upload.parseRequest(request);
```

Comment:
What is the reason to use wildcard without 'extends'?

Reply:
The reason is a copy/paste :-), and I thought it was not possible to fix the type because of the return type of ` parseRequest, but I was wrong.
I changed it with <FileItem>.